crates/turborepo-lib/src/process/mod.rs

```diff
@@ -0,0 +1,286 @@
+}
+
+impl ProcessManager {
+    pub fn new() -> Self {
+        ProcessManager {
+            is_closing: Arc::new(AtomicBool::new(false)),
+            children: Arc::new(Mutex::new(Vec::new())),
+        }
+    }
+}
+
+impl ProcessManager {
+    /// Spawn a new child process to run the given command.
+    ///
+    /// The handle of the child can be either waited or stopped by the caller,
+    /// as well as the entire process manager.
+    ///
+    /// If spawn returns None,
+    pub fn spawn(&self, command: child::Command, stop_timeout: Duration) -> Option<child::Child> {
+        if self.is_closing.load(std::sync::atomic::Ordering::Relaxed) {
+            return None;
+        }
+        let child = child::Child::spawn(command, child::ShutdownStyle::Graceful(stop_timeout));
+        {
+            let mut lock = self.children.lock().unwrap();
+            lock.push(child.clone());
+        }
+        Some(child)
+    }
+
+    /// Stop the process manager, closing all child processes. On posix
+    /// systems this will send a SIGINT, and on windows it will just kill
+    /// the process immediately.
+    pub async fn stop(&self) {
+        self.close(|mut c| async move { c.stop().await }).await
+    }
+
+    /// Stop the process manager, waiting for all child processes to exit.
+    ///
+    /// If you want to set a timeout, use `tokio::time::timeout` and
+    /// `Self::stop` if the timeout elapses.
+    pub async fn wait(&self) {
+        self.close(|mut c| async move { c.wait().await }).await
+    }
+
+    /// Close the process manager, running the given callback on each child
+    ///
+    /// note: this is designed to be called multiple times, ie calling close
+    /// with two different strategies will propagate both signals to the child
+    /// processes. clearing the task queue and re-enabling spawning are both
+    /// idempotent operations
+    async fn close<F, C>(&self, callback: F)
+    where
+        F: Fn(Child) -> C + Sync + Send + Copy + 'static,
+        C: Future<Output = Option<ChildExit>> + Sync + Send + 'static,
+    {
+        self.is_closing
+            .store(true, std::sync::atomic::Ordering::Relaxed);
+
+        let mut set = JoinSet::new();
+
+        {
+            let lock = self.children.lock().expect("not poisoned");
+            for child in lock.iter() {
+                let child = child.clone();
+                set.spawn(async move { callback(child).await });
+            }
+        }
+
+        debug!("waiting for {} processes to exit", set.len());
+
+        while let Some(out) = set.join_next().await {
+            trace!("process exited: {:?}", out);
+        }
+
+        {
+            // clean up the vec and re-open
+            let mut lock = self.children.lock().expect("not poisoned");
+            // just allocate a new vec rather than clearing the old one
+            *lock = vec![];
+            self.is_closing
+                .store(false, std::sync::atomic::Ordering::Relaxed);
+        }
```

Comment:
Just flagging this because I always have a bad time reasoning about atomic orderings. Do we want this to be relaxed?

> Relaxed accesses are the absolute weakest. They can be freely re-ordered and provide no happens-before relationship.

Does this mean compiler/hardware is free to re-order the "set true" and "set false" operations? If so, that's not what we want.

Reading through the [cpp spec 🤢 ](https://en.cppreference.com/w/cpp/atomic/memory_order#Relaxed_ordering) makes it seem that even on the same thread these can get reordered:

> In particular, this may occur if D is completed before C in thread 2, either due to compiler reordering or at runtime